projects/wave-core-new/src/lib/operators/operator-type.model.ts

```diff
@@ -0,0 +1,131 @@
+// export interface OperatorTypeConfig {} // tslint:disable-line:no-empty-interface
+
+/**
+ * Options allowed when cloning the operator
+ */
+export interface OperatorTypeCloneOptions {
+} // tslint:disable-line:no-empty-interface
+
+/**
+ * Dictionary for querying the server.
+ */
+export interface OperatorTypeMappingDict {
+} // tslint:disable-line:no-empty-interface
+
+/**
+ * Dictionary for serializing the operator type.
+ */
+export interface OperatorTypeDict {
+    operatorType: string;
+}
+
+/**
+ * Interface required for complex operator parameter options.
+ * It represents display values of type options.
+ */
+export interface OptionsDict {
+    displayValue: string;
+}
+
+/**
+ * The possible types of parameter values
+ */
+export type ParameterValue = number | string | OptionsDict;
+
+/**
+ * The operator basic type.
+ */
+export abstract class OperatorType {
+
+    /**
+     * Each operator type must have a method that returns
+     * an icon as a data uri image. This provides a default
+     * out of the operator name.
+     */
+    public static createIconDataUrl(iconName: string) {
+        // TODO: replace with proper icons
+        // from `http://stackoverflow.com/questions/3426404/
+        // create-a-hexadecimal-colour-based-on-a-string-with-javascript`
+        const hashCode = (str: string) => { // java String#hashCode
+            let hash = 0;
+            for (let i = 0; i < str.length; i++) {
+                hash = str.charCodeAt(i) + ((hash << 5) - hash); // tslint:disable-line:no-bitwise
+            }
+            return hash;
+        };
+        const intToRGB = (i: number) => {
+            const c = (i & 0x00FFFFFF).toString(16).toUpperCase(); // tslint:disable-line:no-bitwise
+
+            return '00000'.substring(0, 6 - c.length) + c;
+        };
+
+        const color = '#' + intToRGB(hashCode(iconName));
+        const size = 64;
+
+        const canvas = document.createElement('canvas');
+        canvas.width = size;
+        canvas.height = size;
+        const context = canvas.getContext('2d');
+        context.fillStyle = color;
+        context.fillRect(0, 0, 64, 64);
+        return canvas.toDataURL('image/png');
+    }
+
+    /**
+     * Get the server-side name of the type.
+     */
+    abstract getMappingName(): string;
```

Comment:
We should use the chance to refactor this to use readonly public variables that we just override.
This should also work for icon URL.
This prevents creating new strings on each function call but just setting a string once.
The syntax wasn't available back then.

Reply:
1187fb5a6148657ae9eb1e960c54ed287d841f02